atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java

```diff
@@ -165,6 +167,21 @@ protected boolean shouldStopRetrying(int numTimesFailed) {
     public TransactionAndImmutableTsLock setupRunTaskWithConditionThrowOnConflict(PreCommitCondition condition) {
         StartIdentifiedAtlasDbTransactionResponse transactionResponse
                 = timelockService.startIdentifiedAtlasDbTransaction();
+        return wrap(condition, transactionResponse);
+    }
+
+    @Override
+    public List<TransactionAndImmutableTsLock> setupRunTaskWithConditionThrowOnConflictBatch(
+            List<PreCommitCondition> conditions) {
+        List<StartIdentifiedAtlasDbTransactionResponse> transactionResponse
+                = timelockService.startIdentifiedAtlasDbTransactionBatch(conditions.size());
+
+        return Streams.zip(transactionResponse.stream(), conditions.stream(),
+                (response, condition) -> wrap(condition, response)).collect(Collectors.toList());
+    }
+
+    private TransactionAndImmutableTsLock wrap(PreCommitCondition condition,
```

Comment:
You are not dealing with any exception here correctly. If #wrap fails for one of the transactions in the batch, you'll leak everything.